### PR TITLE
Store groups as json

### PR DESCRIPTION
**Pull request by janisozaur**

### CODE REVIEW

File: src/network/network.cpp [hunk 1]
```diff
@@ -260,6 +261,43 @@ void NetworkGroup::Write(NetworkPacket& packet)
...
+	json_t * jsonGroup = json_object();
+	json_object_set_new(jsonGroup, "id", json_integer(id));
+	json_object_set_new(jsonGroup, "name", json_string(GetName().c_str()));
+	json_t * actionsArray = json_array();
+	for (size_t i = 0; i < actions_allowed.size(); i++) {
+		json_array_append_new(actionsArray, json_integer(actions_allowed[i]));
+	}
+	json_object_set_new(jsonGroup, "permissions", actionsArray);
+	return jsonGroup;
+}
```
Comment (IntelOrca) [inline src/network/network.cpp]: I think the serialisation of the permissions should be improved to something like:
```json
"permissions": [
    "chat",
    "terraform",
    "setWaterLevel",
    "clearLandscape"
]
```

Comment (janisozaur) [inline src/network/network.cpp]: I like your suggestion, I'll work on that today.

I'm going to go with latter option:
```
"permissions": [
    "PERMISSION_CHAT",
    "PERMISSION_TERRAFORM",
    "PERMISSION_SET_WATER_LEVEL",
    "PERMISSION_CLEAR_LANDSCAPE"
]
```

**Comment by janisozaur** [comment]
it looks like this now:
```
{
    "default_group": 1,
    "groups": [
        {
            "id": 0,
            "name": "Admin",
            "permissions": [
                "PERMISSION_CHAT",
                "PERMISSION_TERRAFORM",
                "PERMISSION_SET_WATER_LEVEL",
                "PERMISSION_TOGGLE_PAUSE",
                "PERMISSION_CREATE_RIDE",
                "PERMISSION_REMOVE_RIDE",
                "PERMISSION_BUILD_RIDE",
                "PERMISSION_RIDE_PROPERTIES",
                "PERMISSION_SCENERY",
                "PERMISSION_PATH",
                "PERMISSION_CLEAR_LANDSCAPE",
                "PERMISSION_GUEST",
                "PERMISSION_STAFF",
                "PERMISSION_PARK_PROPERTIES",
                "PERMISSION_PARK_FUNDING",
                "PERMISSION_KICK_PLAYER",
                "PERMISSION_MODIFY_GROUPS",
                "PERMISSION_SET_PLAYER_GROUP",
                "PERMISSION_CHEAT",
                "PERMISSION_TOGGLE_SCENERY_CLUSTER"
            ]
        },
        {
            "id": 1,
            "name": "Spectator",
            "permissions": [
                "PERMISSION_CHAT",
                "PERMISSION_PARK_FUNDING"
            ]
        },
        {
            "id": 2,
            "name": "User",
            "permissions": [
                "PERMISSION_CHAT",
                "PERMISSION_TERRAFORM",
                "PERMISSION_SET_WATER_LEVEL",
                "PERMISSION_TOGGLE_PAUSE",
                "PERMISSION_CREATE_RIDE",
                "PERMISSION_REMOVE_RIDE",
                "PERMISSION_BUILD_RIDE",
                "PERMISSION_RIDE_PROPERTIES",
                "PERMISSION_SCENERY",
                "PERMISSION_PATH",
                "PERMISSION_CLEAR_LANDSCAPE",
                "PERMISSION_GUEST",
                "PERMISSION_STAFF",
                "PERMISSION_PARK_PROPERTIES",
                "PERMISSION_PARK_FUNDING",
                "PERMISSION_MODIFY_GROUPS",
                "PERMISSION_SET_PLAYER_GROUP",
                "PERMISSION_TOGGLE_SCENERY_CLUSTER"
            ]
        }
    ]
}
```

**Comment by janisozaur** [comment]
I've done some formatting of `network.cpp` to break ultra-long lines, let me know if you like it.

**Comment by IntelOrca** [comment]
@janisozaur yes fine, is this ready now?

**Comment by janisozaur** [comment]
yes, though you can still give it a spin just to check if there are no glaring issues.

File: src/network/network.cpp [hunk 2]
```diff
@@ -260,6 +275,47 @@ void NetworkGroup::Write(NetworkPacket& packet)
...
+	}
+	group.id = (uint8)json_integer_value(jsonId);
+	group.name = std::string(json_string_value(jsonName));
+	for (size_t i = 0; i < group.actions_allowed.size(); i++) {
+		group.actions_allowed[i] = 0;
+	}
+	for (size_t i = 0; i < json_array_size(jsonPermissions); i++) {
+		json_t * jsonPermissionValue = json_array_get(jsonPermissions, i);
+		const char * perm_name = json_string_value(jsonPermissionValue);
+		int action_id = gNetworkActions.FindCommandByPermissionName(perm_name);
```
Comment (IntelOrca) [inline src/network/network.cpp]: Crashes when perm_name is nullptr

Comment (janisozaur) [inline src/network/network.cpp]: I thought converting `nullptr` to `std::string` creates an empty string?

Comment (IntelOrca) [inline src/network/network.cpp]: You have defined it to take a string pointer, not a string.

Comment (janisozaur) [inline src/network/network.cpp]:
```
int NetworkActions::FindCommandByPermissionName(const std::string &permission_name)
```
It gets converted to `std::string` when passed as an argument.

**Comment by janisozaur** [comment]
should be fixed now

**Pull request merged**